Adds remaining CRUD methods to data store

hannahtinkler opened the pull request:
### Changes proposed in this Pull Request:

Focus: Woo-driven Push Notifications
Team: Apps Infrastructure
Relates to: AINFRA-1435

Apps Infrastructure are currently working with Kiwi to allow WooCommerce instances to trigger their own push notifications, rather than relying on Jetpack Sync. This change adds the remaining methods to the data store that will facilitate CRUD for push tokens. This will be used in the controller.

**Key requirements:**
* A push token can be created, updated, read and deleted
* Push token meta can be created, updated, read and deleted
* A push token can be retrieved using user ID, platform, origin and EITHER token or device UUID

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:
There are no functional changes to test in this PR, but to ensure no side effects, you can:
* Check out this branch to your WooCommerce install
* Ensure your WooCommerce install functions as expected, e.g.:
  * Add a product to cart
  * View WooCommerce settings in WP Admin
  * View WooCommerce orders in WP Admin
* Make an authenticated request to the API to ensure this works as expected, e.g.:
  * https://YOUR_URL/wp-json/wc-admin/features
  * https://YOUR_URL/wp-json/wc/v3/settings
  * https://YOUR_URL/wp-json/wc/v3/customers

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Adds internal library foundations which will eventually be used to send push notifications to user devices. This will support WooCommerce in directly triggering push notifications, instead of relying on Jetpack Sync.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->
This is an internal, non-user facing change.

</details>

hannahtinkler committed:
Adds PushTokenDataStore for push token CRUD operations.

hannahtinkler committed:
Updates handling of updating a token with null device ID.

hannahtinkler committed:
Only delete meta once update is successful.

hannahtinkler committed:
Removes superfluous comments.

hannahtinkler committed:
Minor comment change to retrigger checks.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -190,4 +220,200 @@ public function read_meta( &$push_token ) {
+				LIMIT 1",
+				...$params
+			)
+		);
```
Maybe you can speak a bit more about the expected number of tokens a user will have? As currently written, this query seems less than ideal with these many JOINs. Depending on how things are expected to work, maybe a 2 step approach where we first fetch some candidate post IDs from `post_meta` and then match against the `posts` table could work better.

Or, if this is going to be super high traffic, maybe we'll need a dedicated table.

hannahtinkler replied:
In WPCOM (which is where we'll be moving push tokens from for Woo) 72% of users have 1 token, and 99.85% of users have less than 10. We do have 104 outliers with more than 100 tokens, with a one user having as many as 15135 - these are all for unique devices with multiple registered per day, so are likely to be for test builds. The maximum tokens any user has for WooCommerce apps right now is 394.

The Woo apps will register tokens in the following cases...
* On login
* Periodically every 7 days
* When the app is switched to the foreground **if** it didn't already register a token
* When Firebase regenerates the token for whatever reason

...so I don't expect the endpoint to get a tonne of traffic. I've split the logic out as you suggested, but happy to discuss a dedicated table - what do you think is best?

jorgeatorres reviewed (commented):
Hi @hannahtinkler 👋 

Thanks for your work here!

I know this is work in progress, but with some crucial parts of the architecture missing, the review seems more like a checklist item than useful. The testing instructions are also unrelated to the changes here.

A few comments:

- Overall the code looks is well structured and of high quality. Even though it's unclear how to test, it seems correct to me.
- Not introduced in this PR, but I see that the feature was registered with `skip_compatibility_checks => true` [here](https://github.com/woocommerce/woocommerce/blob/b066fdf8376784ebbfaa3e42119028c5243ac381/plugins/woocommerce/src/Internal/Features/FeaturesController.php#L558). Not sure if that's because you were using the other features as inspiration, but that shouldn't be necessary given default compatibility is "compatible". That is, unless you have a good reason for that.
- `PushToken` should probably inherit from `WC_Data` if the intention is for it to be used similar to other WC objects and as the other datastores for orders, products, etc. work. That might or might not be the intention, but then we should check whether implementing `WC_Object_Data_Store_Interface` makes sense, because it expects `WC_Data` objects to be passed to its methods ([example](https://github.com/woocommerce/woocommerce/blob/ef015f740f89c35cebf34bc0c39a971be6723b68/plugins/woocommerce/includes/interfaces/class-wc-object-data-store-interface.php#L18)). Using `WC_Data` might simplify some of the metadata operations as well.
- Why is the registration of the custom post type not already included? That would allow for more useful testing.

Please let me know what you think!

hannahtinkler committed:
Merge push notifications code with optimizations and tests.

- Add META_KEYS validation and meta_objects_to_map() helper
- Optimize get_by_token_or_device_id() to use simple posts query
- Add null device_uuid deletion in update() method
- Add @since annotations to data store methods
- Simplify array_filter in create/update methods

hannahtinkler committed:
Add push token post type registration.

- Add static register_post_type() method to PushTokensDataStore
- Register post type on init hook from PushNotifications::register()
- Configure post type as private with delete_with_user enabled

hannahtinkler commented:
> I know this is work in progress, but with some crucial parts of the architecture missing, the review seems more like a checklist item than useful. The testing instructions are also unrelated to the changes here.

As these are supporting classes for the device registration endpoint, there's no easy way to test them without adding the endpoint code (which would take this PR to well over 1000 lines). The testing steps described in the PR are more smoke tests to ensure there's no unexpected side effects, and I have some [in-depth testing steps](https://github.com/woocommerce/woocommerce/pull/61384) for the PR that adds the controller that will tests all of this as a byproduct.

> Not introduced in this PR, but I see that the feature was registered with skip_compatibility_checks => true [here](https://github.com/woocommerce/woocommerce/blob/b066fdf8376784ebbfaa3e42119028c5243ac381/plugins/woocommerce/src/Internal/Features/FeaturesController.php#L558). Not sure if that's because you were using the other features as inspiration, but that shouldn't be necessary given default compatibility is "compatible". 

Yep, I was using the other examples as a template - thanks for clarifying, I've updated this! 😄

> PushToken should probably inherit from WC_Data if the intention is for it to be used similar to other WC objects 

That makes sense, added!

> Why is the registration of the custom post type not already included? That would allow for more useful testing.

It was an oversight, thanks for flagging! Have added this 👌

hannahtinkler committed:
Ensures parent constructor is called, and updates for linting errors.

hannahtinkler committed:
Improve test_it_can_add_meta to verify actual insertion.

- Create browser token without device_uuid (optional for browsers)
- Test adding device_uuid meta that does not exist initially
- Verify add_meta actually inserts new row instead of no-op on existing key

hannahtinkler committed:
Removes set_id which is already declared on the parent, and updates tests.

hannahtinkler committed:
Adds missing @since annotations.

hannahtinkler committed:
Update origin to protected, and use data store from container.

hannahtinkler committed:
Remove get_id from PushToken, which is already inherited from WC_Data.

hannahtinkler committed:
Implements code rabbit suggestions.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -13,13 +13,62 @@
+	 * @since 10.5.0
+	 * @internal
+	 */
+	public static function register_post_type() {
```
This is very non-standard for a datastore. I'd suggest the registration of the post type is moved to the `PushNotifications` class that controls the "feature".

hannahtinkler replied:
Makes sense - I moved this 🙂

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -185,9 +265,236 @@ public function read_meta( &$push_token ) {
+	 * @return int|false Meta ID on success, false on failure.
+	 * @throws InvalidArgumentException If the token can't be read or meta key not given.
+	 */
+	public function add_meta( &$push_token, $meta ) {
```
The contract from `WC_Object_Data_Store_Interface::add_meta()` is that `$meta`  is an object (with `key` and `value` properties), not an array, which is how it is treated here.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -185,9 +265,236 @@ public function read_meta( &$push_token ) {
+	 * @return int|bool Meta ID if the key didn't exist, true on successful update, false on failure.
+	 * @throws InvalidArgumentException If the token can't be read or meta key not given.
+	 */
+	public function update_meta( &$push_token, $meta ) {
```
Similarly, `$meta` should be an object with `id`, `key` and `value` props, not an array.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -185,9 +265,236 @@ public function read_meta( &$push_token ) {
+	 * @return bool True on success, false on failure.
+	 * @throws InvalidArgumentException If the token can't be read or meta key not given.
+	 */
+	public function delete_meta( &$push_token, $meta ) {
```
Again, this doesn't respect the contract from `WC_Object_Data_Store_Interface::delete_meta()`.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -185,9 +265,236 @@ public function read_meta( &$push_token ) {
+	 * @param array $meta_objects Array of WC_Meta_Data objects from get_meta_data().
+	 * @return array Associative array with meta keys as keys.
+	 */
+	private function meta_objects_to_map( array $meta_objects ): array {
```
Could you clarify what the goal of this function is? If you need to extract specific metadata from a `WC_Data` object you can always use `get_meta()`.

hannahtinkler replied:
It's been removed now, but it was to convert an array of WC_Meta_Data objects to key => value arrays (for easier access of specific keys).

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -89,15 +136,15 @@ public function read( &$push_token ) {
 		}
 
-		$meta = $this->read_meta( $push_token );
+		$meta_map = $this->meta_objects_to_map( $push_token->get_meta_data() );
```
If the idea is to follow `WC_Data` conventions, `read()` in the datastore method will actually populate the `PushToken` object, not the other way around. What's the point of using `get_meta_data()` on the token itself. Can you clarify?

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -185,9 +265,236 @@ public function read_meta( &$push_token ) {
-			get_post_meta( $push_token->get_id() )
+		global $wpdb;
+
+		$placeholders = implode( ',', array_fill( 0, count( static::META_KEYS ), '%s' ) );
```
`read_meta()` shouldn't hardcode the possible meta keys, but instead let `WC_Data` manage them. Defining internal meta keys makes sense, though, via `get_internal_meta_keys()`, and those will be considered entity properties even when stored in metadata.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -168,6 +213,41 @@ public function update( &$push_token ) {
+	 *
+	 * phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter.FoundAfterLastUsed
+	 */
+	public function delete( &$push_token, $args = array() ) {
```
`WC_Object_Data_Store_Interface::delete()` should return bool.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -168,6 +213,41 @@ public function update( &$push_token ) {
+			throw new Exception( 'Push token could not be found.', WP_Http::NOT_FOUND );
+		}
+
+		foreach ( static::META_KEYS as $key ) {
```
Deleting the post should be enough. Why delete the meta first?

hannahtinkler replied:
I think this was some confusion with meta not being deleted (before this was a force delete) - this block has been removed now 🙂

jorgeatorres reviewed (changes requested):
Hi @hannahtinkler!

Apologies for the delay in re-reviewing and thanks for making the updates and responding to my feedback.

My main concern right now is that the code here is trying to adhere to `WC_Data` for the entity and, as before, `WC_Object_Data_Store_Interface` for the datastore, but it's not fully honoring the contracts they define.

As I said, extending/implementing those classes/interfaces is not a hard requirement (even if it makes _certain_ things easier). Not all datastores need to be traditional datastores (like those for orders of products) **but** if we're going to do it, then we need to make sure the entity and its datastore respect the existing API.

Given this is all in the internal namespace, my suggestion would be to make this as simple as possible for now and consider moving this to use the CRUD layer (`WC_Data` and friends) later on. Right now, unless a big chunk of the methods and logic is rewritten, things won't work as expected.

Please me know what you think.

hannahtinkler committed:
Remove WC_Data and interface dependencies from PushToken.

- Remove WC_Data parent class from PushToken entity
- Remove WC_Object_Data_Store_Interface from PushTokensDataStore
- Move post type registration from DataStore to PushNotifications

hannahtinkler committed:
Cleans up some unused use statements.

hannahtinkler committed:
Reinstates PushToken::PLATFORM_IOS as deprecated.

hannahtinkler commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -555,7 +555,7 @@ private function init_feature_definitions(): void {
 				'is_experimental'              => true,
 				'disable_ui'                   => true,
-				'skip_compatibility_checks'    => true,
+				'skip_compatibility_checks'    => false,
```
Given the feature is marked as `FeaturePluginCompatibility::COMPATIBLE`, this change shouldn't affect any behaviour.

hannahtinkler commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -555,7 +555,7 @@ private function init_feature_definitions(): void {
 				'is_experimental'              => true,
 				'disable_ui'                   => true,
-				'skip_compatibility_checks'    => true,
+				'skip_compatibility_checks'    => false,
```
Given the feature is marked as FeaturePluginCompatibility::COMPATIBLE, this change shouldn't affect any behaviour.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -176,7 +284,7 @@ public function update( &$push_token ) {
 	 * @throws InvalidArgumentException If the token can't be read.
 	 */
-	public function read_meta( &$push_token ) {
+	private function get_meta( PushToken &$push_token ) {
```
Not yet used in core and is an internal method.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -176,7 +284,7 @@ public function update( &$push_token ) {
 	 * @throws InvalidArgumentException If the token can't be read.
 	 */
-	public function read_meta( &$push_token ) {
+	private function get_meta( PushToken &$push_token ) {
```
Not yet used in Core and is internal method.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -176,7 +284,7 @@ public function update( &$push_token ) {
 	 * @throws InvalidArgumentException If the token can't be read.
 	 */
-	public function read_meta( &$push_token ) {
+	private function get_meta( PushToken &$push_token ) {
```
Not yet used in Core and is internal method.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -34,6 +34,13 @@ class PushToken {
+	 *
+	 * @deprecated 10.5.0 Use PLATFORM_APPLE instead
+	 */
+	const PLATFORM_IOS = self::PLATFORM_APPLE;
```
iOS is not a valid platform, this would be incorrect.

hannahtinkler committed:
Reinstates private visibility on PushToken properties.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -28,7 +35,7 @@ class PushTokensDataStore {
 	 * @throws Exception If the token creation fails.
 	 */
-	public function create( &$push_token ) {
+	public function create( PushToken &$push_token ) {
```
Not yet used in Core and is internal method.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -28,7 +35,7 @@ class PushTokensDataStore {
 	 * @throws Exception If the token creation fails.
 	 */
-	public function create( &$push_token ) {
+	public function create( PushToken &$push_token ) {
```
Not yet used in Core and is internal method.

hannahtinkler committed:
Adds void return type hint.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -63,7 +70,7 @@ class PushToken {
 	 */
 	const PLATFORMS = array(
-		self::PLATFORM_IOS,
+		self::PLATFORM_APPLE,
```
There is no usage of this in Core yet to add push tokens, and no external path to add them, there will be no push tokens of type `ios` in the database.

hannahtinkler committed:
Updates PHPdocs for meta methods, and extracts meta query build.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -34,6 +34,13 @@ class PushToken {
+	 *
+	 * @deprecated 10.5.0 Use PLATFORM_APPLE instead
+	 */
+	const PLATFORM_IOS = self::PLATFORM_APPLE;
```
There is no usage of this in Core yet to add push tokens, and no external path to add them, there will be no push tokens of type ios in the database.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -89,7 +94,7 @@ public function read( &$push_token ) {
 		}
 
-		$meta = $this->read_meta( $push_token );
+		$meta = $this->get_meta( $push_token );
```
This is not yet called in Core and is an internal method.

hannahtinkler committed:
Updates post type name with `wc_` prefix.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/PushNotifications.php:
```diff
@@ -53,9 +54,41 @@ public function register(): void {
 			return;
 		}
 
+		add_action( 'init', array( $this, 'register_post_types' ) );
```
Updated to `wc_push_token`

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -63,7 +70,7 @@ class PushToken {
 	 */
 	const PLATFORMS = array(
-		self::PLATFORM_IOS,
+		self::PLATFORM_APPLE,
```
`ios` is not a valid platform so this would not be correct. This is not yet in use, there should be no tokens anywhere with platform `ios`

hannahtinkler committed:
Updates meta method names for clarity.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -34,6 +34,13 @@ class PushToken {
+	 *
+	 * @deprecated 10.5.0 Use PLATFORM_APPLE instead
+	 */
+	const PLATFORM_IOS = self::PLATFORM_APPLE;
```
Interested as to whether you think this is needed @jorgeatorres - Claude suggested it for backward compatibility. The `PLATFORM_IOS` const was added in  10.4.0 (original value `ios`) - it's not used in Core yet and realistically it's unlikely an external developer would have integrated with it, but they _could_ have (although it's an internal library obviously).

jorgeatorres replied:
It's in the internal namespace and this feature hasn't been released yet. I'm ok with not adding a deprecated const. At this stage I'd say you're free to design this however you see fit 😄

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -20,6 +20,13 @@
  * @since 10.5.0
  */
 class PushTokensDataStore {
+	const SUPPORTED_META = array(
+		'origin',
+		'device_uuid',
+		'token',
+		'platform',
+	);
```
"array_map() reindexes the result numerically," -  this is only true if more than one array is passed in as subsequent arguments, which is not true here. This code retains the array keys.

jorgeatorres commented on plugins/woocommerce/tests/php/src/Internal/PushNotifications/PushNotificationsTest.php:
```diff
@@ -175,6 +176,63 @@ public function test_it_caches_enablement_state_correctly() {
+		$this->setExpectedIncorrectUsage( 'WP_Block_Type_Registry::register' );
+
+		// phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
+		do_action( 'init' );
```
Calling `do_action( 'init' )` as part of the test will probably trigger initialization of a lot more things than we want, which can have a lot of side effects.

I'd suggest we just check that `PushNotifications` has registered its `register_post_types()` method by using `has_action( 'init', ... )` and then we just call it to ensure the feature was enabled (or not) depending on the conditions.

(This also applies to the other instances of `do_action( 'init' )`).

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -169,14 +158,120 @@ public function update( &$push_token ) {
+				PushToken::POST_TYPE,
+				$push_token->get_user_id()
+			)
+		);
```
I'm still unclear on exactly when or how tokens are queried, so in the end we might need a dedicated table or something different, but right now, this query could be written using `WP_Query` instead of calling `$wpdb` directly.

Maybe it's even possible to use `meta_query` to query with the meta directly, but if you do decide to keep the 2 steps querying, then it'd greatly help if we first call `update_meta_cache()` with the candidate IDs, which would result in much faster `get_post_meta()` later on.

What do you think?

- - - -

Also, for future reference: if we're not expecting a ton of tokens per user, we could at least cache the IDs of a user's tokens in usermeta, for example.

hannahtinkler replied:
So this method will be called by the create endpoint in the (future) push token controller. This endpoint will be called when the WooCommerce apps need to register a new token, which is typically:
* On user login
* Also once every 7 days.
* Also when the app comes to the foreground if a token hasn't been registered yet
* Also when the app gets a new token from the Firebase service (sometimes it regenerates existing tokens)

When the app makes the call to register a new token, they send the device UUID in the request, and we look up any existing tokens with either the same token value or device UUID - this is because:
* The app doesn't keep track of whether it has already registered a given token, so it just sends it to the server when one of the above happens - this means the server may receive the same token value multiple times, but should only store it once
* There will only ever be one active token per device, so if we get a new token value for a device in our token registry, we should update that existing token because its current token value is now invalid (the app does try to unregister tokens when they become invalid, but this isn't always possible, which is how we may end up getting requests to save a new token for an existing device)

The other time we would query tokens is when sending push notifications, at which point we:
- Query for admin users and `shop_managers`
- Query tokens where the post author is any of those users 

(Of course this usecase wouldn't use `get_by_token_or_device_id`.)

---

I'm totally open to adding a new table - originally I wasn't sure what the best approach would be. CPTs felt more 'WordPressy' but it's starting to feel like it might not be future proof. I don't want to have to worry about migrating the data later if we find the performance isn't where it needs to be - what's your take?

jorgeatorres replied:
I agree that CPTs seem like a good idea initially, and I understand your concern re: migrating later, but using custom tables has some other implications. While we can trust the posts and postmeta tables are always present, altering the database is subject to permissions, etc. so there are a few more scenarios to consider.

That said, it shouldn't prevent us from making the right call here, but unless we run some benchmarks and find that the current approach is wrong, it's hard to tell.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
Given this is not at the response layer, I'd suggest we just throw the exception without the internal server error code, and do something similar for the other exceptions in the datastore.

hannahtinkler replied:
Makes sense, I've removed the `WP_Http` status codes and added `PushTokenNotFoundException` so I'll be able to map the exception type to a 404 response code in the controller 🙂

jorgeatorres reviewed (changes requested):
Hey @hannahtinkler!

Thanks again for making the changes. I agree this looks much cleaner without the CRUD layer, at least at this point.

I left some comments/suggestions, but we're almost there 🐈.

Thank you for your patience!

hannahtinkler committed:
Refactor post type registration tests to avoid side effects.

- Replace do_action('init') with has_action() and direct method calls

hannahtinkler committed:
Add PushTokenNotFoundException for not found errors.

- Introduce PushTokenNotFoundException for cleaner error handling
- Replace generic Exception with PushTokenNotFoundException in data store
- Remove HTTP status codes from internal exceptions

hannahtinkler committed:
Removes other WP_Http usages from data store.

hannahtinkler committed:
Removes deprecated const.

hannahtinkler commented:
Thanks @jorgeatorres 😄 I've actioned your suggestions, and am interested to hear how you think I should handle that query given the extra context!

hannahtinkler committed:
Formatting updates for phpstan?

hannahtinkler committed:
Update function to `use` variable.

hannahtinkler committed:
Update for phpstan.

hannahtinkler committed:
Update typehint.

hannahtinkler committed:
Updates PHPstan baseline.

hannahtinkler committed:
Fixes merge issue resulting in browser-specific logic being removed.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -169,25 +149,172 @@ public function update( &$push_token ) {
+			throw new PushTokenNotFoundException( 'Push token could not be found.' );
+		}
+
+		wp_delete_post( (int) $push_token->get_id(), true );
```
The return value from `wp_delete_post()` is not checked here. Given we're mostly using exceptions, I'm not sure it makes much sense to return a bool in this function, but if we're going to do it, it should reflect whether the post was deleted or not.

hannahtinkler replied:
This was mostly me trying to shut up PHPStan and forgetting I could use the `void` return typehint 😄 I've done this instead

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -169,25 +149,172 @@ public function update( &$push_token ) {
+				PushToken::POST_TYPE,
+				$push_token->get_user_id()
+			)
+		);
```
I believe I mentioned this before, but just in case, unless we're doing a special query that will be much faster, I think this should just be a regular `WP_Query`. All the fields here seem query-able using that API.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -169,25 +149,172 @@ public function update( &$push_token ) {
+		foreach ( $posts as $post ) {
+			$candidate = new PushToken();
+			$candidate->set_id( (int) $post->ID );
+
```
Even if the expected number of tokens here is small, I believe it'd be much faster to perform another `WP_Query`  here if necessary. You can build the meta part using `meta_query` and use `post__in` to limit the query to the IDs found with the first query.

Let me know if I'm missing something about the `$meta` array that makes it not suitable for `meta_query`!

As I mentioned before, if we do need to iterate over the posts and fetch postmeta, we can at least "prime" the cache with `update_meta_cache()` using the post IDs.

hannahtinkler replied:
I don't know if I'm understanding correctly - you [previously commented](https://github.com/woocommerce/woocommerce/pull/62024#discussion_r2547428947) that doing a join for each meta item (as the original query did) would be rough performance-wise, but using `WP_Query` to query posts with the post IDs from the first query and filtering results using `meta_query` appears to do the same thing - if I dump the query being performed by `WP_Query`, there's a join for each piece of meta.

It seems like the most performant approach might be to do a query with `$wpdb` to get all meta for the post IDs returned in the first query (and then use that to filter the tokens in PHP) but I can't find a way to achieve that with `WP_Query`. Let me know if I'm on the wrong track here 😕

jorgeatorres replied:
Yes, I still think one big query that both fetches the posts associated to the user and their metadata could be problematic because of all the JOINs that were originally present, but a second query to fetch all the metadata _only_ for those post IDs already obtained in the first query is probably better than iterating over all the items and running `get_post_meta()` for each one of those.

This query can happen entirely in the `postmeta` table limiting the `post_id` column to the result above and without having to do any JOINs. You can fetch _all_ the metadata associated to those post IDs (which should only be token-related anyways) and filter as necessary instead of going token by token.

The alternative I'm also mentioning is just priming the metadata with `update_meta_cache()` which receives an array of post IDs. That way, the `get_post_meta()` that happens for each token wouldn't hit the database as the data is already cached.

Does that make sense?

hannahtinkler replied:
That makes sense, I've added the call to `update_meta_cache` - thanks for helping me understand how `update_meta_cache` works!

jorgeatorres commented on plugins/woocommerce/phpstan-baseline.neon:
Can you clarify why these suppressions were needed? Can we not rework the code for safety so they are not needed?

hannahtinkler replied:
I didn't realise these were suppressions - it seems like they were added automatically when PHPStan was configured, and got updated after I was prompted to regenerate the baselines by the failure log. 

I've removed all PushNotification entries from the file and resolved the issues, although you may see a lot of redundant-looking type casting and type checks now 😅 These were required to stop a lot of the PHPStan errors.

jorgeatorres reviewed (changes requested):
Hey @hannahtinkler!

Thanks for making some of the changes. I left one last set of comments, though nothing too critical. Let me know what you think!

hannahtinkler committed:
Updates code for PHPStan compliance and removes supressions.

hannahtinkler committed:
Return the correct value from build_meta_array_from_database.

hannahtinkler committed:
Updates the posts query to use WP_Query.

hannahtinkler committed:
Fixes last PHPStan suppression.

hannahtinkler committed:
Ensure correct variable is used in loop.

jorgeatorres reviewed (changes requested):
Hey @hannahtinkler!

This is looking good. With all the bot comments, this PR is getting hard to read, but I responded to your query question [here](https://github.com/woocommerce/woocommerce/pull/62024#discussion_r2618966175).

There are also some failing checks, including PHPStan and related conflicts. Can you take a look at those?
By the way, you can run `composer phpstan:baseline` if the file is out of date with your recent changes, to regenerate the baseline.

hannahtinkler committed:
Adds usage of `update_meta_cache`.

hannahtinkler committed:
Updates phpstan baseline.

hannahtinkler committed:
Updates phpstan baseline.

hannahtinkler commented on plugins/woocommerce/phpstan-baseline.neon:
```diff
@@ -75481,83 +75481,17 @@ parameters:
 
 		-
-			message: '#^Parameter \#1 \$postarr of function wp_update_post expects array\{ID\?\: int, post_author\?\: int, post_date\?\: string, post_date_gmt\?\: string, post_content\?\: string, post_content_filtered\?\: string, post_title\?\: string, post_excerpt\?\: string, \.\.\.\}, array\{ID\: int\|null, post_author\: int\|null, post_type\: ''push_token'', post_status\: ''private'', meta_input\: arra ...
+			message: '#^Parameter \#2 \$object_ids of function update_meta_cache expects array\<int\>\|string, array\<int\|WP_Post\> given\.$#'
```
I can't find a way around these. If I map over $post_ids and cast to an int, I get a different suppression message about not being able to cast `<int\|WP_Post\>` to an int. If I remove `'fields' => 'ids',` and map over the results to return the ID, I get a suppression message about not being able to access property ID on `<int\|WP_Post\>`.

hannahtinkler commented:
Thanks @jorgeatorres - I resolved the check failures and added the meta preload, would you mind taking another look please? 🙂

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -17,12 +17,12 @@ class PushToken {
+	 * Platform identifier for Apple devices.
 	 */
-	const PLATFORM_IOS = 'ios';
+	const PLATFORM_APPLE = 'apple';
```
This is fine as this code lives in the Internal namespace, where there are no guarantees of backwards compatibility.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -17,12 +17,12 @@ class PushToken {
 	 * WordPress post type for storing push tokens.
 	 */
-	const POST_TYPE = 'push_token';
+	const POST_TYPE = 'wc_push_token';
```
This is fine as this code lives in the Internal namespace, where there are no guarantees of backwards compatibility.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -10,55 +10,53 @@
+	 * @return void
 	 */
-	public function create( &$push_token ) {
+	public function create( PushToken &$push_token ): void {
```
This is fine as this code lives in the Internal namespace, where there are no guarantees of backwards compatibility.

jorgeatorres commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -70,26 +68,23 @@ public function create( &$push_token ) {
+	 * @return void
 	 */
-	public function read( &$push_token ) {
+	public function read( PushToken &$push_token ): void {
```
This is fine as this code lives in the Internal namespace, where there are no guarantees of backwards compatibility.

jorgeatorres reviewed (approved):
LGTM. Thank you for all your work on this and for your patience @hannahtinkler!

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -188,10 +188,11 @@ public static function get_new_instance(
 	 * @since 10.4.0
 	 */
-	public function set_id( int $id ) {
+	public function set_id( int $id ): void {
```
Duplicate comment, addressed in other replies.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -70,26 +68,23 @@ public function create( &$push_token ) {
+		$meta = $this->build_meta_array_from_database( $push_token );
 
 		if (
 			empty( $meta['token'] )
```
Duplicate comment, addressed in other replies.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php:
```diff
@@ -10,55 +10,53 @@
+	 * @return void
 	 */
-	public function create( &$push_token ) {
+	public function create( PushToken &$push_token ): void {
```
Duplicate comment, addressed in other replies.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -17,12 +17,12 @@ class PushToken {
+	 * Platform identifier for Apple devices.
 	 */
-	const PLATFORM_IOS = 'ios';
+	const PLATFORM_APPLE = 'apple';
```
Duplicate comment, addressed in other replies.

hannahtinkler commented on plugins/woocommerce/src/Internal/PushNotifications/Entities/PushToken.php:
```diff
@@ -17,12 +17,12 @@ class PushToken {
 	 * WordPress post type for storing push tokens.
 	 */
-	const POST_TYPE = 'push_token';
+	const POST_TYPE = 'wc_push_token';
```
Duplicate comment, addressed in other replies.
